src/Framework/LazyFormattedBuildEventArgs.cs

```diff
@@ -17,33 +16,34 @@ public class LazyFormattedBuildEventArgs : BuildEventArgs
         /// <summary>
         /// Stores the message arguments.
         /// </summary>
-        private object[] arguments;
+        private volatile object argumentsOrFormattedMessage;
```

Comment:
I think this might be an overkill - let's confirm whether we're already being called under a lock anyway.

Reply:
> ```cs
> var args = _arguments;
> if (args != null) {
>   args = CompareAndExchange(ref _arguments, null, args);
>   if (args != null)
>      base.Message = FormateMe(args);
>    return base.Message
> }
> ```

This doesn't quite work. We need all threads calling this code to get the formatted message. In your snippet there is a race window between the winning thread getting non-null result from `CompareAndExchange` and storing the formatted message to `base.Message`. During this time other threads would be getting the original unformatted message.

Reply:
> I think this might be an overkill - let's confirm whether we're already being called under a lock anyway.

Not having to worry about thread safety would be ideal!

Reply:
> In your snippet there is a race window
You are right. I think it is impossible to implement non-blocking `exactly once`. We have to stick to idempotency. 
If  we put both unformated string and arguments into `LazyFormattedBuildEventArgs` we can idempotently build formated message.
```C#
public override string Message
{
    get
    {
        if (RawMessage == null)
        {
            RawMessage = FormatString(_unformated, _args);
        }

        return RawMessage;
    }
}
```
This approach is already used in many EventArgs derives like for example `ProjectStartedEventArgs` or `TaskFinishedEventArgs`